Added conversion from PHP values to DB values during query preparation

petr-buchin opened the pull request:
<!-- Fill in the relevant information below to help triage your pull request. -->

|      Q       |   A
|------------- | -----------
| Type         | feature/improvement
| BC Break     | no
| Fixed issues | #2103 

#### Summary

This PR adds conversion of PHP values in queries to their DB representations using types functionality, and thus allows to query by php values rather that db values.

alcaeus commented:
At a first glance this looks good, thanks! Can you please add some tests for this feature? I'd love to have this in 2.1!

petr-buchin commented:
@alcaeus thanks for the review! Ok, I will add tests this week!

alcaeus commented:
Ignore that for now please - I just haven't found the time to introduce that test to the tip of my boot. It just needs a good kicking here and there. I'll try to fix it later today.

petr-buchin commented:
@alcaeus Thanks to the simple test I added for this feature, I fixed a bug.

If you want me to cover some other cases with tests, just tell me what cases I should consider.

Thanks!

alcaeus commented on lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php:
```diff
@@ -1021,9 +1021,24 @@ public function prepareQueryOrNewObj(array $query, bool $isNewObj = false) : arr
+                if ($this->class->hasField($key)) {
+                    $mapping  = $this->class->fieldMappings[$key];
+                    $typeName = $mapping['type'];
+                    if (Type::hasType($typeName) && ! in_array($typeName, ['collection', 'hash'])) {
```
I would suggest throwing an exception if we can't find the type: this should've been caught during the metadata mapping phase and not be silently ignored here.

petr-buchin replied:
Done

alcaeus commented on tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php:
```diff
@@ -176,6 +183,46 @@ public function testPrepareQueryOrNewObjWithHashIdAndInOperators($hashId)
     }
 
+    /**
+     * @dataProvider provideCustomIds
```
Why does this need a data provider with three identifiers?

petr-buchin replied:
I removed this data provider and replaced it by the data provider you've suggested below

alcaeus commented on tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php:
```diff
@@ -176,6 +183,46 @@ public function testPrepareQueryOrNewObjWithHashIdAndInOperators($hashId)
+
+        $customId = DocumentPersisterCustomTypedId::fromString($objectIdString);
+
+        $value    = ['id' => $customId];
```
Please also add tests for the following queries:
* `['id' => ['$ne' => $customId]]`
* `['id' => ['$in' => [$customId1, $customId2]]]`

I suggest using a data provider for this:
```php
public static function queryProvider() : Generator
{
    yield 'Direct comparison' => ['query' => ['id' => $customId], 'expected' => ['_id' => ...]];
    yield 'Operator with single value' => ...;
}
```

petr-buchin replied:
Done

petr-buchin commented:
Hi @alcaeus, thanks for the review!
> I think we also need to apply the same transformation to arrays
Do you mean we need to convert arrays the same way as we do with a single values?

If yes, how do you want to do this?
From what I see, there is no way to specify item type for a `collection` mapping (except reference collections), thus there is no way to know the type of an item in an array.

petr-buchin commented:
@alcaeus I got your point: I updated tests as you requested and added exception for non-existent types.

I also found and fixed the bug: at the moment ODM user can query by providing target document for reference field like that:

```
['referenceField' => $targetDocumentValue]
```

but user cannot query by providing target document with a query operators, like that:
```
['referenceField' => ['$ne' => $targetDocumentValue]]
```

For now ODM will try to convert target document value in query using target document's id type.
I fixed this behavior by passing retrieving id from target document (see changes in `DocumentPersister::prepareQueryExpression()` method).

alcaeus commented on lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php:
```diff
@@ -1021,15 +1021,59 @@ public function prepareQueryOrNewObj(array $query, bool $isNewObj = false) : arr
+     *
+     * @return mixed
+     */
+    private function convertToDatabaseValue($fieldName, $value)
```
For new code, please use type declarations where possible:

```suggestion
    private function convertToDatabaseValue(string $fieldName, $value)
```

alcaeus commented on lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php:
```diff
@@ -1021,15 +1021,59 @@ public function prepareQueryOrNewObj(array $query, bool $isNewObj = false) : arr
+    /**
+     * Converts a single value to its database representation based on the mapping type
+     *
+     * @param string $fieldName
```
Obsolete due to the type declaration requested below.
```suggestion
```

petr-buchin replied:
Done.

alcaeus commented on lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php:
```diff
@@ -1021,15 +1021,59 @@ public function prepareQueryOrNewObj(array $query, bool $isNewObj = false) : arr
+     * Converts a single value to its database representation based on the mapping type
+     *
+     * @param string $fieldName
+     * @param mixed  $value
```
```suggestion
     * @param mixed $value
```

alcaeus commented on lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php:
```diff
@@ -1244,6 +1288,11 @@ private function prepareQueryExpression(array $expression, ClassMetadata $class)
             // Process query operators whose argument arrays need type conversion
             if (in_array($k, ['$in', '$nin', '$all']) && is_array($v)) {
                 foreach ($v as $k2 => $v2) {
+                    if ($v2 instanceof $class->name) {
```
I assume that this is the bugfix that you mentioned in your comment:

> For now ODM will try to convert target document value in query using target document's id type.
I fixed this behavior by retrieving id from target document (see changes in `DocumentPersister::prepareQueryExpression()` method).

Can you please add a comment as to why this conditional is necessary? This code is already quite loosely commented, so any comment explaining why we're doing things will help when we will eventually have to touch this piece of code again. Thanks!

petr-buchin replied:
Done.

alcaeus commented on tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php:
```diff
@@ -176,6 +184,70 @@ public function testPrepareQueryOrNewObjWithHashIdAndInOperators($hashId)
+     *
+     * @dataProvider queryProviderForCustomTypeId
+     */
+    public function testPrepareQueryOrNewObjWithCustomTypedId(array $testCase)
```
```suggestion
    public function testPrepareQueryOrNewObjWithCustomTypedId(array $expected, array $query)
```

petr-buchin replied:
Done.

alcaeus commented on tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php:
```diff
@@ -176,6 +184,70 @@ public function testPrepareQueryOrNewObjWithHashIdAndInOperators($hashId)
+        Type::registerType('DocumentPersisterCustomId', DocumentPersisterCustomIdType::class);
+
+        $this->assertEquals(
+            $testCase['expected'],
```
```suggestion
            $expected,
```

alcaeus commented on tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php:
```diff
@@ -176,6 +184,70 @@ public function testPrepareQueryOrNewObjWithHashIdAndInOperators($hashId)
+
+        $this->assertEquals(
+            $testCase['expected'],
+            $documentPersister->prepareQueryOrNewObj($testCase['query'])
```
```suggestion
            $documentPersister->prepareQueryOrNewObj($query)
```

alcaeus commented on tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php:
```diff
@@ -176,6 +184,70 @@ public function testPrepareQueryOrNewObjWithHashIdAndInOperators($hashId)
+            $customId2
+        );
+
+        // We need to have test cases inside this method, since we cannot access `$this->dm` in data provider
```
We can work around this by using a closure for the `getReference` logic above. Consider this short example:
```php
public function dataProvider(): array
{
    return [[function (DocumentManager $dm) { return $dm->getReference(...); }]];
}

/** @dataProvider dataProvider */
public function testCase(Closure $getReference)
{
    $reference = $getReference($this->dm);
    // ...
}
```

petr-buchin replied:
I've refactored this test using your advice, but I'm not sure that I understood you correctly.
Please look at the code and confirm if that's what you wanted.

The downside of closure is that we don't get `$expected` and `$query` as arguments of the test method as in the test above.

alcaeus commented on tests/Doctrine/ODM/MongoDB/Tests/Functional/DocumentPersisterTest.php:
```diff
@@ -223,6 +295,36 @@ public function testPrepareQueryOrNewObjWithSimpleReferenceToTargetDocumentWithN
+
+        yield 'Direct comparison' => [
+            [
+                'query' => ['id' => $customId],
```
With the suggestion above, you can drop one array level:
```php
yield 'Direct comparison' => [
    'query' => ...,
    'expected' => ...,
];
```

The key names for "query" and "expected" aren't necessary since PHPUnit passes the values of the array as arguments to the test. However, they make working with the data provider easier since you don't need to think which array is which and instead can use the parameter names.

petr-buchin replied:
Done, but only for the test method that does not use closure for `getReference()` logic

petr-buchin commented:
@alcaeus thanks for the suggestions!
I will fix this later today

petr-buchin commented:
@alcaeus sorry about not getting back with fixes sooner.

I did all the fixes as you've requested, but I'm not sure about one single fix: please confirm if fix with `getReference()` logic, data provider and closure is what you wanted.

Thanks!

alcaeus commented:
Thanks @petr-buchin!

petr-buchin commented:
@alcaeus thanks for review and for help with this!
